presto-kinesis/src/main/java/io/prestosql/plugin/kinesis/KinesisRecordSet.java

```diff
@@ -0,0 +1,494 @@
+        {
+            checkArgument(field < columnHandles.size(), "Invalid field index");
+            return columnHandles.get(field).getType();
+        }
+
+        /**
+         * Advances the cursor by one position, retrieving more records from Kinesis if needed.
+         * <p>
+         * We retrieve records from Kinesis in batches, using the getRecordsRequest.  After a
+         * getRecordsRequest we keep iterating through that list of records until we run out.  Then
+         * we will get another batch unless we've hit the limit or have caught up.
+         *
+         * @return
+         */
+        @Override
+        public boolean advanceNextPosition()
+        {
+            if (shardIterator == null && getRecordsRequest == null) {
+                getIterator(); // first shard iterator
+                log.info("Starting read.  Retrieved first shard iterator from AWS Kinesis.");
+            }
+
+            if (getRecordsRequest == null || (!listIterator.hasNext() && shouldGetMoreRecords())) {
+                getKinesisRecords();
+            }
+
+            if (listIterator.hasNext()) {
+                return nextRow();
+            }
+            else {
+                log.info("Read all of the records from the shard:  %d batches and %d messages and %d total bytes.", batchesRead, totalMessages, totalBytes);
+                return false;
+            }
+        }
+
+        /**
+         * Determine whether or not to retrieve another batch of records from Kinesis.
+         */
+        private boolean shouldGetMoreRecords()
+        {
+            return shardIterator != null && batchesRead < maxBatches &&
+                    getMillisBehindLatest() > MILLIS_BEHIND_LIMIT;
+        }
+
+        /**
+         * Retrieves the next batch of records from Kinesis using the shard iterator.
+         * <p>
+         * Most of the time this results in one getRecords call.  However we allow for
+         * a call to return an empty list, and we'll try again if we are far enough
+         * away from the latest record.
+         */
+        private void getKinesisRecords()
+                throws ResourceNotFoundException
+        {
+            // Normally this loop will execute once, but we have to allow for the odd Kinesis
+            // behavior, per the docs:
+            // A single call to getRecords might return an empty record list, even when the shard contains
+            // more records at later sequence numbers
+            boolean fetchedRecords = false;
+            int attempts = 0;
+            while (!fetchedRecords && attempts < fetchAttempts) {
+                long now = System.currentTimeMillis();
+                if (now - lastReadTime <= sleepTime) {
+                    try {
+                        Thread.sleep(now - lastReadTime);
+                    }
+                    catch (InterruptedException e) {
+                        log.error("Sleep interrupted.", e);
```

Comment:
This swallows the thread interrupted status. Instead, this should
```java
Thread.currentThread().interrupted();
throw new RuntimeException("thread interrupted");
```
See https://www.ibm.com/developerworks/library/j-jtp05236/index.html

Reply:
changed